### PR TITLE
rgw: S3 lifecycle now supports expiration date

### PR DESCRIPTION
**Pull request by zhangsw**
Signed-off-by: Zhang Shaowen <zhangshaowen@cmss.chinamobile.com>

### CODE REVIEW

**Comment by zhangsw** [comment]
The s3-test is here https://github.com/ceph/s3-tests/pull/170

File: src/rgw/rgw_lc.cc [hunk 1]
```diff
@@ -92,6 +96,19 @@ int RGWLifecycleConfiguration::check_and_add_rule(LCRule *rule)
   return 0;
 }
 
+bool RGWLifecycleConfiguration::has_same_action(const lc_op& first, const lc_op& second) {
+  if ((first.expiration > 0 || first.expiration_date != boost::none) &&
```
Comment (dang) [inline src/rgw/rgw_lc.cc]: I think a better solution is to add an "empty()" or "valid()" method to LCExpiration that checks date and days.  That way all the checks are simple for all the rule types.  If "empty()" is used, then all the checks above can stay the same.

This also means that the code checking validity doesn't need to change when we add new types in the future.

File: src/rgw/rgw_lc.h [hunk 2]
```diff
@@ -59,8 +66,15 @@ class LCExpiration
...
-    return days.empty();
+  bool has_days() const {
+    return !days.empty();
+  }
+  void set_date(const string& _date) { date = _date; }
+  string get_date() const{
+    return date;
+  }
+  bool has_date() const {
+    return !date.empty();
```
Comment (dang) [inline src/rgw/rgw_lc.h]: Should check for boost::none as well.

Comment (zhangsw) [inline src/rgw/rgw_lc.h]: date here is a string

File: src/rgw/rgw_lc.cc (hunk 1 above)
Comment (zhangsw) [inline src/rgw/rgw_lc.cc]: I've updated the pr as you advised

**Comment by yuriw** [comment]
http://pulpito.ceph.com/yuriw-2017-07-12_19:35:54-rgw-wip-yuri-testing_2017_7_13-distro-basic-smithi/

**Comment by cbodley** [comment]
jenkins test this please

**Pull request merged**